### PR TITLE
TEST: test coverage for sklearnex SPMD ifaces

### PR DESCRIPTION
**Pull request by ethanglaser**
# Description
Scope:
 - Introduces initial validation of spmd algos via mpi-pytest
 - Identify and create follow-up tasks for issues identified in initial testing

For each sklearnex spmd algorithm, individual manually-created tests and parametrized synthetic tests validate model attributes and prediction results against batch implementations

Known issues to address in follow-up:
 - KMeans centers not aligned after init
 - Number of KMeans iterations do not align
 - PCA falls back if mle or n_components integer
 - PCA fails if n_rows_rank < n_cols < n_rows
 - Forest results are not aligned
 - LinReg fails if n_rows_rank < n_cols < n_rows
 - LogReg model coefficients do not align
 - LogReg fails if n_rows_rank < n_cols < n_rows
 - kNN needs all classes represented on each process initially

### CODE REVIEW

File: sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py
Comment (samir-nasibli) [inline sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py:38]: better to move into try section above

File: sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py
Comment (KulikovNikita) [inline sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py:21]: Please, make it a common functionality to be used in other tests

File: sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py
Comment (KulikovNikita) [inline sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py:43]: What about moving this code template to a common test fixture?

**Comment by ethanglaser** [comment]
Job with infra branch: http://intel-ci.intel.com/eefc46ad-f4dd-f157-a13b-a4bf010d0e2e

**Comment by ethanglaser** [comment]
Latest CI with infra branch: http://intel-ci.intel.com/ef0c0adc-f849-f14c-a086-a4bf010d0e2e

File: onedal/tests/utils/_spmd_support.py
Comment (samir-nasibli) [inline onedal/tests/utils/_spmd_support.py:88]:
```suggestion
def generate_clustering_data(n_samples, n_features, centers=None, random_state=42):
    X, y = make_blobs(
        n_samples=n_samples, centers=centers, n_features=n_features, random_state=random_state
    )
    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=random_state)
    return X_train, X_test, y_train, y_test
```

Let's have `random_state` as a param for the generators

Comment (ethanglaser) [reply onedal/tests/utils/_spmd_support.py]: done

**Comment by ethanglaser** [comment]
There was mention of using https://github.com/intel/scikit-learn-intelex/blob/main/onedal/cluster/tests/test_dbscan.py#L52 logic for dbscan synthetic data generation, but on closer look its just a basic random data generation, not using epsilon or any dbscan params. Since I have been able to generate meaningful data using make_blobs, I am going to stick with it unless there are objections.

**Review by samir-nasibli — COMMENTED** [review]
Initially reviewed.

File: onedal/tests/utils/_spmd_support.py
Comment (ethanglaser) [reply onedal/tests/utils/_spmd_support.py:21]: Good point, I originally put sklearn and just changed it yesterday but will change it back

File: sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py
Comment (samir-nasibli) [inline sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py:21]: Better to use relative imports inside module. Please fix it for the rest of the files here

Comment (ethanglaser) [reply sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py]: fixed

File: onedal/tests/utils/_spmd_support.py
Comment (ethanglaser) [reply onedal/tests/utils/_spmd_support.py]: I'm all for reusing functionality but not sure that its necessary in this case, it just increases complexity of the code here

**Comment by ethanglaser** [comment]
Latest CI: http://intel-ci.intel.com/ef389422-f0b4-f19a-9e6d-a4bf010d0e2e

File: sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py
Comment (ethanglaser) [reply sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py]: Yields `ImportError: attempted relative import with no known parent package`. Returning to non relative import for now, unless there is a suggested workaround that would allow for relative import

**Comment by samir-nasibli** [comment]
@ethanglaser for a such great job done!

I'll take an even deeper look at all the PR later.

I would suggest this testing to be as consistent as possible with other modules in sklearnex.

All SPMD ifaces have interability with dpnp/dpctl inputs. I suggest saving this for testing.
SPMD ifaces are held up
1) DPNP inputs with GPU sycl_queue
2) DPCTL inputs with GPU sycl_queue
3) numpy inputs and queue param with GPU sycl_queue provided.

This will be extended for Array API inputs as well.

I suggest to use `get_dataframes_and_queues(dataframe_filter_="dpnp,dpctl", device_filter_="gpu",)`, `_as_numpy` and `_convert_to_dataframe` like in other algos modules testing.

I am also working on refactoring and extending `get_dataframes_and_queues` for `numpy` inputs and sycl_queue on #1909 
further integration. After #1909 it is just needed updated to `get_dataframes_and_queues(dataframe_filter_="dpnp,dpctl,np_sycl", device_filter_="gpu")` for enabling it for such where `numpy` input with queue param enabled.

**Comment by ethanglaser** [comment]
> I suggest to use `get_dataframes_and_queues(dataframe_filter_="dpnp,dpctl", device_filter_="gpu",)`, `_as_numpy` and `_convert_to_dataframe` like in other algos modules testing.


Thanks, addressed

**Comment by ethanglaser** [comment]
Latest CI with infra branch: http://intel-ci.intel.com/ef3fe1d1-c22d-f1dc-b5ca-a4bf010d0e2e

File: sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py
Comment (olegkkruglov) [reply sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py:25]: due to sklearn guideline imports must be absolute in tests, so no need to look for workaround
https://scikit-learn.org/stable/developers/develop.html

File: sklearnex/tests/_utils_spmd.py
Comment (samir-nasibli) [inline sklearnex/tests/_utils_spmd.py:123]: Better to use `_as_numpy` https://github.com/intel/scikit-learn-intelex/blob/04dbc5e1a412ac76f3578f76508df87e46487910/onedal/tests/utils/_dataframes_support.py#L115C5-L115C14

File: sklearnex/tests/_utils_spmd.py
Comment (samir-nasibli) [inline sklearnex/tests/_utils_spmd.py:179]: please use https://github.com/intel/scikit-learn-intelex/blob/04dbc5e1a412ac76f3578f76508df87e46487910/onedal/tests/utils/_dataframes_support.py#L115C5-L115C14 instead

File: sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py
Comment (olegkkruglov) [inline sklearnex/spmd/basic_statistics/tests/test_basic_statistics_spmd.py:65]: I guess it makes sense to use updated API here

File: onedal/tests/utils/_spmd_support.py
Comment (samir-nasibli) [reply onedal/tests/utils/_spmd_support.py]: Make sense do refactoring separately. Have a draft PR https://github.com/intel/scikit-learn-intelex/pull/1936 , that probably will address this.

**Review by samir-nasibli — COMMENTED** [review]
Please provide some info for failed tests. Maybe make sense disable them, that could be addressed separately

File: sklearnex/tests/_utils_spmd.py
Comment (ethanglaser) [reply sklearnex/tests/_utils_spmd.py:123]: Good catch, meant to remove this previously

File: sklearnex/tests/_utils_spmd.py
Comment (ethanglaser) [reply sklearnex/tests/_utils_spmd.py:143]: Agreed, original was incorrect. Now updated.

**Review by olegkkruglov — APPROVED** [review]
LGTM, assuming green CI. And please don't forget to create tickets for skipped tests before merge.

File: sklearnex/tests/_utils_spmd.py
Comment (olegkkruglov) [inline sklearnex/tests/_utils_spmd.py:87]:
```suggestion
    # Generates statistical data
```

**Comment by olegkkruglov** [comment]
It looks like all tests are skipped in both public and private CI. is it expected?

**Comment by ethanglaser** [comment]
> It looks like all tests are skipped in both public and private CI. is it expected?

They will only run with GPU and dpctl (private CI GPU tests on py 3.9-3.10)

File: sklearnex/tests/_utils_spmd.py
Comment (ethanglaser) [reply sklearnex/tests/_utils_spmd.py]: good catch

**Comment by ethanglaser** [comment]
Latest CI with infra branch: http://intel-ci.intel.com/ef4a161a-e33a-f1fc-b023-a4bf010d0e2e

File: sklearnex/spmd/covariance/tests/test_covariance_spmd.py
Comment (samir-nasibli) [inline sklearnex/spmd/covariance/tests/test_covariance_spmd.py:67]: I see that you are mixing use of numpy's assert_allclose and your own implemented _spmd_assert_allclose. Why not to use it for all places?

**Review by samir-nasibli — COMMENTED** [review]
Overall looks good to me. Thank you!

Two questions:
1) I don't see skips in test suits for the know issues, is it possible define defects by disabling related test cases?
2) does it work when a number of processes to run the test <2?
Does it make sense to add ```@pytest.mark.mpi(min_size=2)```?

**Review by samir-nasibli — COMMENTED** [review]
Also it is not clear to me how you run this tests?
https://pytest-mpi.readthedocs.io/en/latest/usage.html#

**Comment by olegkkruglov** [comment]
I see also that different dtypes are not tested here. Is this parametrizing skipped on purpose?

**Comment by ethanglaser** [comment]
> I see also that different dtypes are not tested here. Is this parametrizing skipped on purpose?

Not on purpose - I will add it in. Good point.

**Comment by ethanglaser** [comment]
> Also it is not clear how you run this tests? https://pytest-mpi.readthedocs.io/en/latest/usage.html#
> 
> I think make sense to add some readme files/update docs in case how to run this tests.

Currently only added to internal CI since this is where we have GPU validation - see infra PR 712. Not sure if it makes sense to document in public repo.

File: sklearnex/spmd/covariance/tests/test_covariance_spmd.py
Comment (ethanglaser) [reply sklearnex/spmd/covariance/tests/test_covariance_spmd.py:67]: Certain attributes and results will require it and others do not. For example with linear regression, inference is done on chunks of data with spmd (whereas batch does entire data) therefore spmd version is used whereas coefficients are universal and require regular assert_allclose

**Comment by ethanglaser** [comment]
> 1. I don't see skips in test suits for the know issues, is it possible define defects by disabling related test cases?

There are pytest.skip( ) within the test functions - mostly because skips are dependent on parameters to function (except forest)

> does it work when a number of processes to run the test <2?
>  Does it make sense to add @pytest.mark.mpi(min_size=2)?

Infra sets up tests to run with 4 processes, but it would just run batch GPU if 1 is used, so I don't see why it would not work. But all tests are set up to skip if mpi or gpu are not present

**Comment by samir-nasibli** [comment]
> Currently only added to internal CI since this is where we have GPU validation - see infra PR 712. Not sure if it makes sense to document in public repo.

We are adding a testing module to the public repository. It seems to me it makes sense to add a doc with info about how to launch, including some links to mpi-pytest is given.
Could be done in this PR or in the separate PR as well. We have a check box in our PR templates for adding/updatung docs.

**Review by samir-nasibli — APPROVED** [review]
Looks good to me. Just minor comments, that could be probably addressed in the separate PRs.
Assuming green CI

**Comment by ethanglaser** [comment]
> Yes, I saw it before. My question raised because it is not everywhere. For example, for Kmeans I did not find the skip for an issue specified in the description.

True - in some cases its not handled by pytest.skip() but instead aspects of the test are commented. KMeans for instance the iter check is commented out and init check, because I want to have some validation still. If there is a good way to handle this instead with pytest.skip I am open to other ideas.

> We are adding a testing module to the public repository. It seems to me it makes sense to add a doc with info about how to launch, including some links to mpi-pytest is given. Could be done in this PR or in the separate PR as well. We have a check box in our PR templates for adding/updatung docs.

I will update SPMD docs updates task to include this - we currently do not have documentation of our spmd interfaces at all so I don't know if it could be added yet.

Thanks for reviews. Still working out some CI fixes with introduction of float32 but will share job once its cleaner.

**Comment by samir-nasibli** [comment]
I see some test fails on the latest CI job provided.
I will not dismiss my approve, but please attach the green CI link before the merge. Feel free to ask review again if required.

File: sklearnex/tests/_utils_spmd.py
Comment (olegkkruglov) [inline sklearnex/tests/_utils_spmd.py:131]:
```suggestion
    assert_allclose(_as_numpy(spmd_result), _as_numpy(local_batch_result), **kwargs)
```
it allows to compare spmd results with results of batch estimator computed on GPU. comparison with GPU results makes a little bit more sense because CPU and GPU results are slightly different and it forces to increase tolerance.

Comment (ethanglaser) [reply sklearnex/tests/_utils_spmd.py:131]: Original decision from arch meeting was to compare to batch CPU estimators, which are numpy by default

Comment (olegkkruglov) [reply sklearnex/tests/_utils_spmd.py:131]: suggested change still allows to compare to batch CPU estimators but also adds ability to compare to GPU results. is there any reason you want to make comparison to GPU forbidden here?

**Pull request merged**
